rust/connlib/tunnel/src/lib.rs

```diff
@@ -110,53 +77,201 @@ pub(crate) fn get_v6(ip: IpAddr) -> Option<Ipv6Addr> {
+            peers_by_id: HashMap::new(),
+        }
+    }
+
+    fn handle_socket_packet<'a>(
+        &'a mut self,
+        packet: io::Result<Received<'a>>,
+    ) -> Option<device_channel::Packet<'a>> {
+        match packet {
+            Ok(Received {
+                local,
+                from,
+                packet,
+            }) => {
+                match self.node.decapsulate(
+                    local,
+                    from,
+                    packet,
+                    std::time::Instant::now(),
+                    self.write_buf.as_mut(),
+                ) {
+                    Ok(Some((conn_id, packet))) => {
+                        tracing::trace!(target: "wire", %local, %from, bytes = %packet.packet().len(), "read new packet");
+                        let Some(peer) = self.peers_by_id.get(&conn_id) else {
+                            tracing::error!(%conn_id, %local, %from, "Couldn't find connection");
+                            return None;
+                        };
+                        return peer
+                            .untransform(packet.source(), self.write_buf.as_mut())
+                            .ok();
+                    }
+                    Ok(None) => None,
+                    Err(e) => {
+                        tracing::error!(%local, %from, "Failed to decapsulate incoming packet: {e}");
+                        None
+                    }
+                }
+            }
+            Err(e) => {
+                tracing::error!("Failed to read socket: {e:#?}");
+                None
+            }
         }
     }
 }
 
+struct ConnectionState<TRole, TId, TTransform> {
+    connections: Connections<TRole, TId, TTransform>,
+    connection_pool_timeout: BoxFuture<'static, std::time::Instant>,
+    sockets: Sockets,
+}
+
+impl<TRole, TId, TTransform> ConnectionState<TRole, TId, TTransform>
+where
+    TId: Eq + Hash + Copy + fmt::Display,
+    TTransform: PacketTransform,
+{
+    fn new(private_key: StaticSecret) -> Result<Self> {
+        Ok(ConnectionState {
+            connections: Connections::new(Node::new(private_key, std::time::Instant::now())),
+            connection_pool_timeout: sleep_until(std::time::Instant::now()).boxed(),
+            sockets: Sockets::new()?,
+        })
+    }
+
+    fn cleanup(&mut self) -> Vec<TId> {
+        let peers_to_stop: Vec<_> = self
+            .connections
+            .peers_by_id
+            .iter()
+            .filter_map(|(id, _)| {
+                if let Some((_, info)) = self.connections.node.stats().find(|(i, _)| id == i) {
+                    if info.missed_keep_alives() >= 2 {
+                        tracing::debug!(%id, "connection expired");
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                } else {
+                    tracing::debug!(%id, "connection expired");
+                    Some(*id)
```

Comment:
It's good to leave this here for hardening but I notice that sometimes `ConnectionFailed` event is not emitted